src/lib/peer_trust/peer_trust.ml

```diff
@@ -31,30 +30,243 @@ module type S = sig
 
   val create : db_dir:string -> t
 
-  val record : t -> peer -> action -> unit
+  val ban_pipe : t -> peer Strict_pipe.Reader.t
+
+  val record : t -> Logger.t -> peer -> action -> unit Deferred.t
 
   val lookup : t -> peer -> Peer_status.t
 
   val close : t -> unit
 end
 
-module Make (Peer : sig
-  include Hashable.S
+module Make
+    (Peer : Sexpable.S) (Now : sig
+        val now : unit -> Time.t
+    end)
+    (Action : Action_intf)
+    (Db : Key_value_database.S
+          with type key := Peer.t
+           and type value := Record.t) =
+struct
+  type t =
+    { db: Db.t
+    ; bans_reader: Peer.t Strict_pipe.Reader.t
+    ; bans_writer:
+        (Peer.t, Strict_pipe.synchronous, unit Deferred.t) Strict_pipe.Writer.t
+    }
+
+  module Record_inst = Record.Make (Now)
 
-  val sexp_of_t : t -> Sexp.t
-end) (Now : sig
-  val now : unit -> Time.t
-end)
-(Action : Action_intf)
-(Db : Key_value_database.S with type key := Peer.t and type value := Record.t) :
-  S with type peer := Peer.t and type action := Action.t = struct
-  type t = unit
+  let create ~db_dir =
+    let reader, writer = Strict_pipe.create Strict_pipe.Synchronous in
+    {db= Db.create ~directory:db_dir; bans_reader= reader; bans_writer= writer}
 
-  let create = stub ()
+  let ban_pipe {bans_reader} = bans_reader
 
-  let record = stub ()
+  let lookup {db} peer =
+    match Db.get db peer with
+    | Some record -> Record_inst.to_peer_status record
+    | None -> Record_inst.to_peer_status @@ Record_inst.init ()
 
-  let lookup = stub ()
+  let close {db; bans_writer} =
+    Db.close db ;
+    Strict_pipe.Writer.close bans_writer
 
-  let close = stub ()
+  let record {db; bans_writer} log peer action =
+    let log' = Logger.child log "peer_trust" in
+    let old_record =
+      match Db.get db peer with
+      | None -> Record_inst.init ()
+      | Some trust_record -> trust_record
+    in
+    let new_record =
+      match Action.to_trust_response action with
+      | Insta_ban -> Record_inst.ban old_record
+      (* I don't like runtime exceptions, but the trust change constructors
+         should only be called with constant arguments, so I think any bugs
+         that trigger this will be very visible in testing. *)
```

Comment:
Static assertions like this are really useful as you're saying -- we have a `Debug_assert` to help out with this for things that you don't want to check in production (assertions are compiled out). Here it's probably fine, though (you can remove this comment)